describe/describe.go

```diff
@@ -0,0 +1,267 @@
+// Copyright © 2017 thingful
+
+package describe
+
+import (
+	"fmt"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/spf13/cast"
+)
+
+type typez string
+
+const (
+
+	// Int32 type is an signed int32
+	Int32 = typez("int32")
+
+	// Int64 type is an signed int64
+	Int64 = typez("int64")
+
+	// String type is a string
+	String = typez("string")
+
+	// Url type is a value that is a valid URL
+	Url = typez("url")
+
+	// Bool type is a true/false value
+	Bool = typez("bool")
+)
+
+// Parameter describes a configuration parameter
+type Parameter struct {
+	Name        string
+	Type        typez
+	Required    bool
+	Description string
+	Default     interface{}
+	Examples    []string
+}
+
+// Parameters are a collection of Paramater structs
+type Parameters []Parameter
+
+// Describe returns a string description of the Parameter
+func (p Parameter) Describe() string {
+
+	if p.Default == nil {
+		return fmt.Sprintf("%s : %s (Type : %s, Required : %v, Examples %v )", p.Name, p.Description, p.Type, p.Required, p.Examples)
+	}
+
+	return fmt.Sprintf("%s : %s (Type : %s, Required : %v, Default : %v, Examples %v )", p.Name, p.Description, p.Type, p.Required, p.Default, p.Examples)
+}
+
+func NewValues(config map[string]string, params Parameters) (Values, error) {
+
+	values := Values{}
+
+	// validate params
+	for _, p := range params {
+
+		if p.Required {
+
+			v, found := config[p.Name]
+
+			if !found {
+				return values, fmt.Errorf("%s is Required but not supplied", p.Name)
+			}
+			values[p.Name] = Value{Parameter: p, Value: v}
+		} else {
+
+			v, found := config[p.Name]
+
+			if found {
+				values[p.Name] = Value{Parameter: p, Value: v}
+			}
+		}
+	}
+	// validate types
+	for k, v := range values {
+
+		switch v.Type {
+
+		case String:
+
+			_, ok := values.String(k)
+
+			if !ok {
+				return values, fmt.Errorf("%s is not of type String", k)
+			}
+
+		case Int32:
+
+			_, ok := values.Int32(k)
+
+			if !ok {
+				return values, fmt.Errorf("%s is not of type Int32", k)
+			}
+
+		case Int64:
+
+			_, ok := values.Int64(k)
+
+			if !ok {
+				return values, fmt.Errorf("%s is not of type Int64", k)
+			}
+
+		case Bool:
+			_, ok := values.Bool(k)
+
+			if !ok {
+				return values, fmt.Errorf("%s is not of type Bool", k)
+			}
+
+		case Url:
+			_, ok := values.Url(k)
+
+			if !ok {
+				return values, fmt.Errorf("%s is not of type Url", k)
+			}
+
+		default:
+			return values, fmt.Errorf("unknown type : %s", string(v.Type))
+		}
+	}
+
+	return values, nil
+}
+
+// Value contains the Parameter type description alongside its value
+type Value struct {
+	Parameter
+	Value interface{}
+}
+
+// Values are a collection of Value structs
+type Values map[string]Value
+
+// String returns a string and true if a value exists and can be cast to a string
+func (v Values) String(key string) (string, bool) {
+
+	value, found := v[key]
```

Comment:
if eventually describe will be used into a goroutine, it will be needed to modify Values type and add a Mutex to it.